### ISSUE TITLE
Does ramp-workflow have an end-user Python API?

### ISSUE THREAD

**Issue by GaelVaroquaux** [paris-saclay-cds/ramp-workflow#131]
I find it very tedious to work with the shell script "ramp_test_submission", in particular because it is hard to use an interactive debugger.

**Comment by jorisvandenbossche** [comment]
As I completely agree with your sentiment, I typically use the actual model that you write in an interactive session (in my case in a notebook, but that does not matter). Eg some snippets from the mars-craters notebook to make this work (using sys.path manipulation to import things):

```
import problem
X_train, y_train = problem.get_train_data()
X_test, y_test = problem.get_train_data()

import sys
sys.path.append('./submissions/keras_ssd7/')
from object_detector import ObjectDetector

model = ObjectDetector()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
...
```

but this is certainly not yet ideal, and also does not directly give you access to the cross-validation and scoring that `ramp_test_submission` does.

**Comment by glemaitre** [comment]
This is what is executed under the hood I think:

https://github.com/paris-saclay-cds/ramp-workflow/blob/68146005369b31c1c855c2372172d355440994a1/rampwf/utils/testing.py#L63-L142

**Comment by GaelVaroquaux** [comment]
Half an hour of trying not to do this, I give up: I feel that it is a
huge loss of productivity for me to work via shell scripts, and I will
use Joris's trick.

I am a bit worried that this undocumented practice is probably beneficial
for many others.

**Comment by jorisvandenbossche** [comment]
> I am a bit worried that this undocumented practice is probably beneficial for many others.

I used it explicitly in the starting kit notebook for mars craters as example, and I think said before to Guillaume that I personally would do the same for autism. But of course, that is certainly not yet "documented"

**Comment by glemaitre** [comment]
We might use the trick:
```
%load_ext autoreload
%autoreload 2
```

**Comment by jorisvandenbossche** [comment]
> That said, I worry about the import statements that mean that users need to reset the kernel each time (as imports are cached)

For the example in the notebook, we can also actually put the model code in the notebook if it is not too long, then this should not be a problem (for mars-craters this was very long, that's why I did the sys.path import)

**Comment by glemaitre** [comment]
@GaelVaroquaux I think that I can modify the notebook to implement the different classifier and run a `cross_validate` of scikit-learn (depending of the version with a single score).

We would need to pipeline the transformer and classifier define in the notebook I think.

**Comment by glemaitre** [comment]
This would work with minor changes (make `FeatureExtractor` to  be real scikit-learn transformer)
```python
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import cross_validate
from problem import get_cv

pipe = make_pipeline(FeatureExtractor(), Classifier())
cv = get_cv(data_anatomy, labels)
results = cross_validate(pipe, data_anatomy, labels, scoring='roc_auc', cv=cv,
                         verbose=100, return_train_score=True)
print(results)
```

We could even create a function always doing this in the workflow section and calling it redifining the Classifier and FeatureExtractor. @GaelVaroquaux WDYT?

**Comment by glemaitre** [comment]
ok I updated the notebook

**Comment by kegl** [comment]
I would love any improvement along these lines. I did something similar in the Titanic notebook https://github.com/ramp-kits/titanic/blob/master/titanic_starting_kit.ipynb, just scroll down to "Working in the notebook". It's a step-by-step exeuction of what `ramp_test_submission` does on a single fold. Any improvement would be nice, e.g. running the full CV, etc. I think `imp.load_source` solves the reloading problem. The other option is to copy paste `problem.py` in the notebook and modify stuff there, then when it's don, copy-paste it back into the standalon py file. 

In general, problem.py was not optimized for the building phase of RAMP, but @aboucaud is working on a bolierplate to make that easier https://github.com/ramp-kits/cookiecutter-ramp-kit. 

Personally, once problem.py and the starting kit is fixed and debugged, I find `ramp_test_submission` efficient for playing with the models.

**Comment by GaelVaroquaux** [comment]
> Personally, once problem.py and the starting kit is fixed and debugged, I find ramp_test_submission efficient for playing with the models.

No interactive debugger? You're missing out.

**Comment by agramfort** [comment]
I agree. It would be great to have an --pdb option like nose to jump into
pdb when there is an error. I don't know if it's hard to do though.

**Comment by GaelVaroquaux** [comment]
> I agree. It would be great to have an --pdb option like nose to jump into pdb when there is an error. I don't know if it's hard to do though.

IMHO, a first-class citizen Python API would be better: people are more
productive in their favorite Python environment, whether it is IPython,
the notebook, jupyterlab, spyder...

**Comment by kegl** [comment]
Another direction is to modularize stuff in `testing.py` further so functions become callable from the notebook.

**Comment by kegl** [comment]
The main purpose of `ramp_test_submission` is that it is a simple unit test. We didn't have debugging in mind. What it solved was that now we do not have to write a test for each RAMP separately, the only "moving part" is `problem.py`. But there is nothing magical about it and I'm sure you guys can figure out how to make it better. My only specs is that there should be a one-liner command line script that tests the submission files before submission and which is as close as possible to what we do at the backend, to avoid, as much as possible, errors after submission. At this point with @aboucaud we are working on actually running the exact same script at the backend (with the only difference of using a different data set) to eliminate all difference between backend and starting kit.

**Comment by glemaitre** [comment]
The `Estimator` class allows to use pure scikit-learn pipeline and we will be fine now (or almost)
